### ISSUE TITLE
classmethod defined before `__init__` sees incorrect signature

### ISSUE THREAD

**Issue by JelleZijlstra** [python/mypy#3384]
```python
class Foo:
    @classmethod
    def classm(cls, a: int, b: str) -> Foo:
        reveal_type(cls)  # E: Revealed type is 'def () -> classm.Foo'
        return cls(a, b)  # E: Too many arguments for "Foo"

    def __init__(self, a: int, b: str) -> None:
        self.a = a
        self.b = b
```
Apparently, we eagerly bind the cls parameter of a classmethod to the type as it exists when semanal.py visits the classmethod. If that is before `__init__` has been defined, then we incorrectly infer that `cls` takes no arguments. It works as expected (no errors; reveal_type says cls takes two parameters) if I put `classm` after `__init__`.

**Comment by JelleZijlstra** [comment]
This is a dupe of #1727.
